bokeh/settings.py

```diff
@@ -157,250 +128,516 @@
 # Dev API
 #-----------------------------------------------------------------------------
 
-#-----------------------------------------------------------------------------
-# Private API
-#-----------------------------------------------------------------------------
+def convert_str(value):
+    ''' Return a string as-is
+    '''
+    return value
+
+def convert_bool(value):
+    ''' Convert a string to True or False
+
+    If a boolean is passed in, it is returned as-is. Otherwise the function
+    maps the following strings, ignoring case:
+
+    * "yes", "1", "on" -> True
+    * "no", "0", "off" -> False
+
+    Args:
+        value (str):
+            A string value to convert to bool
+
+    Returns:
+        bool
+
+    Raises:
+        ValueError
+
+    '''
+    if value in (True, False):
+        return value
+
+    val = value.lower()
+    if val in ["yes", "1", "on", "true", "True"]:
+        return True
+    if val in ["no", "0", "off", "false", "False"]:
+        return False
+
+    raise ValueError("Cannot convert {} to boolean value".format(value))
+
+def convert_str_seq(value):
+    ''' Convert a string to a lit of strings
+
+    If a list or tuple is passed in, it is returned as-is.
+
+    Args:
+        value (str) :
+            A string to convert to a list of strings
+
+    Returns
+        list[str]
+
+    '''
+    if isinstance(value, (list, tuple)):
+        return value
+
+    try:
+        return value.split(",")
+    except:
+        raise ValueError("Cannot convert {} to list value".format(value))
+
+_log_levels = {
+    "CRITICAL" : logging.CRITICAL,
+    "ERROR"    : logging.ERROR,
+    "WARNING"  : logging.WARNING,
+    "INFO"     : logging.INFO,
+    "DEBUG"    : logging.DEBUG,
+    "TRACE"    : 9,  # Custom level hard-coded to avoid circular import
+    "NONE"     : None,
+}
+
+def convert_logging(value):
+    '''Convert a string to a Python logging level
+
+    If a log level is passed in, it is returned as-is. Otherwise the function
+    understands the following strings, ignoring case:
+
+    * "critical"
+    * "error"
+    * "warning"
+    * "info"
+    * "debug"
+    * "trace"
+    * "none"
+
+    Args:
+        value (str):
+            A string value to convert to a logging level
+
+    Returns:
+        int or None
+
+    Raises:
+        ValueError
+
+    '''
+    if value in set(_log_levels.values()):
+        return value
+
+    value = value.upper()
+    if value in _log_levels:
+        return _log_levels[value]
+
+    raise ValueError("Cannot convert {} to log level, valid values are: {}".format(value, ", ".join(_log_levels)))
+
+class _Unset: pass
+
+def is_dev():
+    return convert_bool(os.environ.get("BOKEH_DEV", False))
+
+class PrioritizedSetting(object):
+    ''' Return a value for a global setting according to configuration precedence.
+
+    The following methods are searched in order for the setting:
+
+    6. immediately supplied values
```

Comment:
So there is no way to override the default? E.g.:

https://github.com/bokeh/bokeh/blob/a20f671f1e297c45ca8d3acc3c7d897cc3c1ec00/bokeh/resources.py#L81

This breaks `BOKEH_RESOURCES` and the entire development workflow, including `test_examples`.

Reply:
In fact I can't make `test_examples` to use anything by `inline`.

Reply:
I figured this out last night. The issue is that there is a cannonically accepted order to search or settings which puts *explicitly passed*  values above env vars, but we have never made a distinction between values explicitly supplied, and default parameter values. The right solution is to do that, by making the default values `None` so that e.g. `settings.resources(mode)` will defer to the the other settings sources. [1] I will do that tonight. That makes the settings the single ground truth for these defaults everywhere and ensures that they are always deduced in the same way everywhere. 


[1] We could also use an explicit `Unset` default value, but at least right now we don't have any cases where `None` is a valid value that would be explicitly passed in, so it works to denote "missing optional"